packages/slice-machine/src/hooks/usePersistedState.ts

```diff
@@ -1,27 +1,40 @@
 import { Dispatch, SetStateAction, useEffect, useState } from "react";
+import { AnyZodObject } from "zod";
 
 type PersistedState<T> = [T, Dispatch<SetStateAction<T>>];
 
 const SLICE_MACHINE_STORAGE_PREFIX = "slice-machine";
 
+type UsePersistedStateOptions = {
+  schema?: AnyZodObject;
+};
+
 export function usePersistedState<T>(
   key: string,
 ): PersistedState<T | undefined>;
 export function usePersistedState<T>(
   key: string,
   defaultValue: T,
+  options?: UsePersistedStateOptions,
 ): PersistedState<T>;
 export function usePersistedState<T>(
   key: string,
   defaultValue?: T,
+  options?: UsePersistedStateOptions,
 ): PersistedState<T | undefined> {
+  const { schema } = options ?? {};
   const computedKey = `${SLICE_MACHINE_STORAGE_PREFIX}_${key}`;
 
   const [value, setValue] = useState<T | undefined>(() => {
     try {
       const value = localStorage.getItem(computedKey);
 
-      return value !== null ? (JSON.parse(value) as T) : defaultValue;
+      if (value == null) return defaultValue;
+      if (!schema) return JSON.parse(value) as T;
+
+      const result = schema.safeParse(JSON.parse(value));
+      if (!result.success) throw result.error;
+      return result.data as T;
```

Comment:
If we're going to throw anyway, let's use `schema.parse`. Also, we shouldn't cast here, as the schema is supposed to help us with that. But I guess with the previous comment this will be solved.

Reply:
Yeah makes total sense, done!